### PR TITLE
Block internal policies from being assigned to tokens

### LINKED ISSUE #1618: Internal policies should not list, and root users should not be able to add them to a token

### PR DESCRIPTION
**Pull request by LLBennett**
DO NOT MERGE. STILL BEING WORKED.

Fixes #1618

### CODE REVIEW

File: vault/policy_store.go [hunk 1]
```diff
@@ -89,7 +92,7 @@ func (c *Core) setupPolicyStore() error {
 	// Ensure that the cubbyhole response wrapping policy exists
 	policy, err = c.policyStore.GetPolicy(cubbyholeResponseWrappingPolicyName)
 	if err != nil {
-		return errwrap.Wrapf("error fetching default policy from store: {{err}}", err)
+		return errwrap.Wrapf("error fetching cubbyhole response wrapping policy from store: {{err}}", err)
```
Comment (vishalnayak) [inline vault/policy_store.go]: Might be better to say "error fetching 'response-wrapping' policy from store:"

Comment (jefferai) [inline vault/policy_store.go]: Agreed...we could change the variable names internally at some point, but the user-facing side of this ended up with just "response wrapping".

File: vault/policy_store.go [hunk 2]
```diff
@@ -114,7 +117,7 @@ func (ps *PolicyStore) SetPolicy(p *Policy) error {
 	if p.Name == "" {
 		return fmt.Errorf("policy name missing")
 	}
-	if strutil.StrListContains(immutablePolicies, p.Name) {
+	if strutil.StrListContains(immutablePolicies, p.Name) || strutil.StrListContains(nonAssignablePolicies, p.Name) {
```
Comment (jefferai) [inline vault/policy_store.go]: This isn't needed as any policy that is non-assignable will also be immutable.

File: vault/policy_store.go [hunk 3]
```diff
@@ -210,13 +213,31 @@ func (ps *PolicyStore) ListPolicies() ([]string, error) {
...
+	}
+
+	return keys, err
 }
 
 // DeletePolicy is used to delete the named policy
 func (ps *PolicyStore) DeletePolicy(name string) error {
 	defer metrics.MeasureSince([]string{"policy", "delete_policy"}, time.Now())
-	if strutil.StrListContains(immutablePolicies, name) {
+	if strutil.StrListContains(immutablePolicies, name) || strutil.StrListContains(nonAssignablePolicies, name) {
```
Comment (jefferai) [inline vault/policy_store.go]: Same here.

File: vault/token_store.go [hunk 4]
```diff
@@ -1193,6 +1193,13 @@ func (ts *TokenStore) handleCreateCommon(
 		return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
 	}
 
+	// Prevent internasl policies from being assigned to tokens
```
Comment (jefferai) [inline vault/token_store.go]: s/internasl/internal

File: vault/token_store.go [hunk 5]
```diff
@@ -1193,6 +1193,13 @@ func (ts *TokenStore) handleCreateCommon(
...
+	for _, policy := range te.Policies {
+		if strutil.StrListContains(nonAssignablePolicies, policy) {
+			return logical.ErrorResponse(fmt.Sprintf("cannot assign %s policy", policy)), nil
```
Comment (jefferai) [inline vault/token_store.go]: I'm on the fence whether the right thing to do is simply filter them out, or return an error. Returning an error does force fixing bad code, but in other places we manipulate the asked-for list (e.g. to add `default`).

File: vault/policy_store.go (hunk 1 above)
Comment (LLBennett) [inline vault/policy_store.go]: Corrected.

File: vault/policy_store.go (hunk 2 above)
Comment (LLBennett) [inline vault/policy_store.go]: Wondered if they might at some point diverge. Either way, it's corrected.

File: vault/policy_store.go (hunk 3 above)
Comment (LLBennett) [inline vault/policy_store.go]: Corrected.

File: vault/token_store.go (hunk 4 above)
Comment (LLBennett) [inline vault/token_store.go]: Fixed!

File: vault/policy_store.go [hunk 6]
```diff
@@ -89,7 +92,7 @@ func (c *Core) setupPolicyStore() error {
 	// Ensure that the cubbyhole response wrapping policy exists
 	policy, err = c.policyStore.GetPolicy(cubbyholeResponseWrappingPolicyName)
 	if err != nil {
-		return errwrap.Wrapf("error fetching default policy from store: {{err}}", err)
+		return errwrap.Wrapf("error fetching response wrapping policy from store: {{err}}", err)
```
Comment (vishalnayak) [inline vault/policy_store.go]: `s/response wrapping/response-wrapping`

**Comment by LLBennett** [comment]
Based on the go ahead from Jeff, merging. Closes hashicorp/vault#1618

**Pull request merged**